src/app/clusters/service-area-server/service-area-server.h

```diff
@@ -0,0 +1,751 @@
+     * @brief Get the Current Location 
+     * @return The current location
+     */
+    DataModel::Nullable<uint32_t> GetCurrentLocation();
+
+    /**
+     * @brief Set the current location
+     * @param aCurrentLocation where the device is currently. Must be a supported location or NULL.
+     * @return true if the current location is set
+     * @note if current location is set to null, estimated end time will be set to null
+     */ 
+    bool SetCurrentLocation(const DataModel::Nullable<uint32_t> & aCurrentLocation);
+
+
+    //*************************************************************************
+    // Estimated End Time manipulators
+
+    /**
+     * @brief Get the estimated end time
+     * @return The estimated epoch time in seconds when operation at the location indicated by the CurrentLocation attribute will be completed, in seconds.
+     */
+    DataModel::Nullable<uint32_t> GetEstimatedEndTime();
+
+    /**
+     * @brief Set the estimated end time
+     * @param aEstimatedEndTime The estimated epoch time in seconds when operation at the location indicated by the CurrentLocation attribute will be completed, in seconds.
+     * @return true if attribute is set
+     * @note change notification is generated (!!) if appropriate per cluster requirements
+     */ 
+    bool SetEstimatedEndTime(const DataModel::Nullable<uint32_t> & aEstimatedEndTime);
+
+
+    //*************************************************************************
+    // Progress list manipulators
+
+    /**
+     * @brief Get the list of current progress element Id's
+     * @param aSupportedLocationIds empty vector to be filled with location Id's
+    */
+    void GetProgressElementIds(std::vector<uint32_t> & aSupportedLocationIds);
+
+    /**
+     * @brief Get a progress element using the locationId
+     * @param aLocationId the locationId to use
+     * @param aProgressElement const pointer to the found map name - NULL if not found
+     * @return true if element is found and returned
+     * @note progress element pointer has limited lifetime - only valid until progress list is changed
+     * @note change progress element contents with provided set functions
+    */
+    bool GetProgressElementById(uint32_t aLocationId, const Structs::ProgressStruct::Type*& aProgressElement);
+
+    /**
+     * @brief Is the progress element in the progress list?
+     * @param aLocationId location id of the progress element
+     * @return true if the progress element identified by Id is in the progress list
+     */ 
+    bool IsProgressElement(uint32_t aLocationId);
+
+    /**
+     * @brief Add a progress element containing inactive status to the progress list
+     * @param aLocationId location id of the progress element
+     * @return true if the new progress element passed validation checks and was successfully added to the list
+     * @note caller is responsible for change notification
+    */
+    bool AddPendingProgressElement(uint32_t aLocationId);
```

Comment:
Choice was to provide specific setters for each field of a Progress element - full element references are always read-only.  This allows validation and possibly action when any field is updated.